rfcs/2022-04-using-native-dialog.md

```diff
@@ -0,0 +1,102 @@
+
+Implementation of this component should not change the current component APIs for `<Drawer>` and `<Dialog>`.
+
+The reasoning behind creating one parent component is as follows:
+
+- Code is maintained in a singular file; DRY implementation
+- The `<dialog>` spec has a limited number of methods available, making the complexity of this component minimal
+  - `close()` will close the dialog
+  - `show()` will open the dialog, but background elements will remain interactive
+  - `showModal()` will open the dialog as a modal, background elements will **not** be interactive
+
+Simplified example of implementation:
+
+```js
+const _Dialog = (props) => {
+  <dialog className={customClass} {...dialogProps}>
+    {children}
+  </dialog>
+}
+
+const Drawer = (props) => {
+  <_Dialog className='ds-c-drawer' hasFocusTrap={bool}>
+    {/* Header */}
+    <Button onClose={handleOnClose}>
+    {/* Body */}
+    {/* Optional footer */}
+  </_Dialog>
+}
+
+const Dialog = (props) => {
+  <_Dialog className='ds-c-dialog'>
+    {/* Header */}
+    <Button onClose={handleOnClose}>
+    {/* Body */}
+    {/* Footer - Modal action buttons */}
+  </_Dialog>
+}
+```
+
+## Benefits
+
+By relying on the browser's implementation, the Design System stands to inherit several benefits:
+
+- Basic functional and accessibility enhancements become the browser's problem
+- Less reliance on third-parties, results in
+  - Publishing less code (better performance)
+  - A more flexible implementation (no longer beholden to someone else's API)
+  - Less Engineering time and effort spent updating
+
+## Risks
+
+Two risks have been identified with this approach:
+
+1. The Design System needs to implement a polyfill to continue IE support, which may result in less bundle size savings than initially suggested
+2. There may be specific implementations for `<Dialog>` and `<Drawer>` that weren't tested for in the initial discovery
+
+### 1. Polyfill support
+
+[Windows support of IE will end on June 15, 2022](https://docs.microsoft.com/en-us/lifecycle/announcements/internet-explorer-11-end-of-support). In alignment with other websites, and to encourage Users to update their browsers, I recommend removing all IE-specific polyfills following this date.
+
+**tldr; The addition of this polyfill is a temporary concern.**
+
+### 2. One-off bugs may be discovered
+
+The discovery of using the `<dialog>` element in the Design System was done using the pre-existing Storybook examples and documentation pages. As a result, there is a chance consuming teams have specific use-cases not documented in our system that this implementation will expose.
+
+## Questions and Requested Feedback
+
+- How does the Design System demarcate internal, not for public-consumption components?
+  - Is prefixing a component with an underscore, as in `<_Dialog>`, acceptable? Or do we prefix the name with something like, `<InternalDialog>`?
+  - Should the system have a special directory for these components?
```

Comment:
I don't mind the `_Dialog` naming convention. The `_` seems like a standard way the industry denotes something that is internal.
I also wouldn't mind having an `Internal` directory that this (and future internal components) live in to really drive the point home.

Reply:
Also, to avoid confusion between our component called `Dialog` and because it will live in a top-level folder as a sibling of `Dialog`, I do think we actually name it something like `NativeDialog`. I don't have a _strong_ opinion, but I think underscored internal things should only exist within a module. I don't think there's any reason why we couldn't expose `NativeDialog` in our package, especially if it could be useful for child design systems.

Reply:
So should the component structure/name look like the following:

```
design-system/
|-- src/
|-- components/
|   |-- Dialog/
|   |-- Internal/
|   |   |-- NativeDialog/
|   |   |   |-- NativeDialog.tsx
```

Reply:
Per Patrick's comment, if we are going to expose `NativeDialog` in our package, then I think it should just live along with the other components, like
```
design-system/
|-- src/
|    |-- components/
|    |    |-- Dialog/
|    |    |-- NativeDialog/
|    |    |    |-- NativeDialog.tsx
```

my original proposal was only if we weren't going to make `NativeDialog` available to consumers